### PR TITLE
[onert] Define LossCode and LossInfo as separated files

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit implements LossCode and LossInfo as separated files and use the same structure in Compiler, TrainingInfo and Loss IR.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

### CODE REVIEW

File: runtime/onert/core/include/ir/operation/Loss.h [hunk 1]
```diff
@@ -36,35 +36,16 @@ class Loss : public Operation
...
-  };
-
 public:
-  Loss(const OperandIndexSequence &inputs, const OperandIndexSequence &outputs, const Param &param);
+  Loss(const OperandIndexSequence &inputs, const OperandIndexSequence &outputs);
 
 public:
   void accept(OperationVisitor &v) const override;
-  std::string name() const override;
+  std::string name() const override { return _loss_name + toString(opcode()); };
```
Comment (ragmani) [inline runtime/onert/core/include/ir/operation/Loss.h:44]: The name seems to be duplicated. Is `_loss_name` necessary?
```suggestion
  std::string name() const override { return toString(opcode()); };
```

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: The `_loss_name` is necessary to print the detailed loss type. The opcode() only prints `Loss`, but `_loss_name` has one of the detailed loss name of `MeanSquaredError` or `CategoricalCrossentropy`.

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: GraphDump calls the operation through ir::Operation visitor to print the loss name. However, the Loss in ir::Operation does not have detailed train information. It is included in ir::train::Operation. Thus, the _loss_name will be updated from the derived class to the base class.

Comment (ragmani) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: I see. How about removing moving this method to `ir::train::operation::Loss`? If so, you can remove `_loss_name` and implement more simply like below.
```suggestion
  std::string name() const override { return toString(_param.loss_code); };
```

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: If this function moves to the `ir::train::operation::Loss`, the GraphDump does not print the `_param.loss_code` because GraphDump uses the `ir::OperationVisitor` not `ir::train::OperationVisitor`... :(

**Review by zetwhite — APPROVED** [review]
LGTM :+1:

File: runtime/onert/core/include/ir/operation/Loss.h (hunk 1 above)
Comment (glistening) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: I think `Loss` is similar to `BinaryArithmetic` in inference operation. Why do we need to define `Loss` as a concrete op? Is there any problem to define `MeanSquaredError` and `CategoricalCrossEntrophy` as operations respectively?

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: I thought about it. Loss operation is only necessary in the train kernel, but onert is designed to operate each operation in `ir::operation`. Therefore, if `LossLayer` is not defined in `ir::operation` and each loss function such as `MeanSquaredError` or `CategoricalCrossentropy` is implemented, this operation should be located in `ir::operation` rather than `ir::train::operation`. If we can implement the operation in `ir::train::operation` without defining in `ir::operation`, your approach seems appropriate.

In this PR, I'd like to introduce `LossCode` and `LossInfo` classes. Therefore, it would be good to discuss it on a separate issue. Thank you for your comment.

Comment (ragmani) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: > If this function moves to the ir::train::operation::Loss, the GraphDump does not print the _param.loss_code because GraphDump uses the ir::OperationVisitor not ir::train::OperationVisitor... :(

Does `ir::OperationVisitor` block overloading virtual functions?

Comment (ragmani) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: > Is there any problem to define MeanSquaredError and CategoricalCrossEntrophy as operations respectively?

As far as I know, the reason why `ir::Operation` that can define a operation among multiple operations such as `BinaryArithmetic` are introduced is duplicated code. I don't know other problems.

Comment (glistening) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: > In this PR, I'd like to introduce `LossCode` and `LossInfo` classes. Therefore, it would be good to discuss it on a separate issue. Thank you for your comment.

Okay. Let's consider this approach later.

Comment (glistening) [reply runtime/onert/core/include/ir/operation/Loss.h:44]: > I thought about it. Loss operation is only necessary in the train kernel, but onert is designed to operate each operation in `ir::operation`. Therefore, if `LossLayer` is not defined in `ir::operation` and each loss function such as `MeanSquaredError` or `CategoricalCrossentropy` is implemented, this operation should be located in `ir::operation` rather than `ir::train::operation`. If we can implement the operation in `ir::train::operation` without defining in `ir::operation`, your approach seems appropriate.

Just a note:

`TrainableOperator` already has its own operator list. - `runtime/onert/core/include/ir/train/Operations.lst`. Thus, at a glance, I don't find problem.

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h]: @glistening @ragmani 
I finally understood you guys suggestion. I thought of implementing the `name()` function in derived classes such as 'MeanSquaredError` or `CategoricalCrossentropy` classes. In this case, I thought it couldn't do that because these classes were derived classes and the operation visitor couldn't call the derived class function. Now I know that `name()` can be implemented in `ir::train::operation::Loss` class, not `ir::operation::Loss` class. I updated this PR. Thank you.

Comment (jyoungyun) [reply runtime/onert/core/include/ir/operation/Loss.h]: There is one think we need to know. The output varies depending on what GraphDumper hands over.

```
[   GraphDumper  ]   %11 = @4_MeanSquaredErrorLoss(%9, %14)
```

**Review by glistening — APPROVED** [review]
LGTM. (I want to try to remove `ir::operation::Loss` after merging this.)

**Review by zetwhite — APPROVED** [review]

**Review by ragmani — APPROVED** [review]
LGTM

**Comment by glistening** [comment]
> LGTM. (I want to try to remove `ir::operation::Loss` after merging this.)

I took a look. It requires more changes than I guessed. It is related how to reuse the existing code, which is based and designed for inference, with less effort.
